### PR TITLE
[Core] Defining bounding box and  `ObjectType` in `DynamicBins`

### PR DESCRIPTION
**Pull request by loumalouomega**
**📝 Description**

Basically what has been done for `Tree` in https://github.com/KratosMultiphysics/Kratos/pull/11730 and https://github.com/KratosMultiphysics/Kratos/pull/11737 (plus https://github.com/KratosMultiphysics/Kratos/pull/11739) but for `DynamicBins`.

- Defining `GetBoundingBox` method in `DynamicBins` class
- Defining `SizeType`, `IndexType`, and `CoordinateType`, ... as aliases with `using`.
- Adding documentation for member functions: `ExistPoint`, `SearchNearestPoint`, `SearchInRadius`, and `SearchInBox`.
- Updating comments and documentation for better code understanding in general.
- Minor code formatting improvements.
- Adding `GetObjectType` and define `ObjectType` in `Tree`. The `ObjectType` is taken from `PointType` when this is defined, otherwise `void` is defined. An auxiliary static method is defined in order to determine when `ObjectType` is defined.

**🆕 Changelog**

- [Defining bounding box and  `ObjectType` in `DynamicBins`](https://github.com/KratosMultiphysics/Kratos/commit/7326911095ce2ad23a77601f2e5edf0306f13458)

### CODE REVIEW

**Comment by loumalouomega** [comment]
@roigcarlo

**Comment by loumalouomega** [comment]
> @roigcarlo

helloooooo

**Comment by roigcarlo** [comment]
> helloooooo

Hi

**Comment by loumalouomega** [comment]
> > helloooooo
> 
> Hi

🤗

File: kratos/spatial_containers/bins_dynamic.h [hunk 1]
```diff
@@ -802,23 +915,28 @@ class BinsDynamic : public TreeNode<TDimension,TPointType, TPointerType, TIterat
     BinsDynamic(BinsDynamic const& rOther);
 
 private:
+    ///@name Member Variables
+    ///@{
+
+    IteratorType mitPointsBegin;         /// Iterator to the first point
+    IteratorType mitPointsEnd;           /// Iterator to the last point
```
Comment (roigcarlo) [inline kratos/spatial_containers/bins_dynamic.h]: Why the change of the name? we have never called iterators `mitXXXX`

File: kratos/spatial_containers/bins_dynamic.h [hunk 2]
```diff
@@ -36,232 +52,303 @@ std::size_t TDimension,
...
-        for(SizeType i = 0 ; i < Dimension ; i++)
-        {
-            mMinPoint[i] = MinPoint[i];
-            mMaxPoint[i] = MaxPoint[i];
+        mNumCells = std::distance(mitPointsBegin, mitPointsEnd);
+        auto& r_min_point = GetMinPoint();
+        auto& r_max_point = GetMaxPoint();
+        for (SizeType i = 0; i < Dimension; i++) {
+            r_min_point[i] = MinPoint[i];
+            r_max_point[i] = MaxPoint[i];
```
Comment (roigcarlo) [inline kratos/spatial_containers/bins_dynamic.h]:
```c++
BinsDynamic(IteratorType const& PointBegin, IteratorType const& PointEnd, PointType const& MinPoint, PointType const& MaxPoint, SizeType BucketSize = 1)
        : mitPointsBegin(PointBegin), mitPointsEnd(PointEnd), mBoundingBox(MinPoint, MaxPoint)
    {
        if (mitPointsBegin == mitPointsEnd)
            return;

        mNumCells = std::distance(mitPointsBegin, mitPointsEnd);
 ```

File: kratos/spatial_containers/bins_dynamic.h [hunk 3]
```diff
@@ -36,232 +52,303 @@ std::size_t TDimension,
...
-        for(SizeType i = 0 ; i < Dimension ; i++)
-        {
-            mMinPoint[i] = MinPoint[i];
-            mMaxPoint[i] = MaxPoint[i];
+        auto& r_min_point = GetMinPoint();
+        auto& r_max_point = GetMaxPoint();
+        for (SizeType i = 0; i < Dimension; i++) {
+            r_min_point[i] = MinPoint[i];
+            r_max_point[i] = MaxPoint[i];
         }
```
Comment (roigcarlo) [inline kratos/spatial_containers/bins_dynamic.h]:
```c++
BinsDynamic(PointType const& MinPoint, PointType const& MaxPoint, SizeType BucketSize)
        : mNumCells(0), mBoundingBox(MinPoint, MaxPoint)
    {
```

File: kratos/spatial_containers/bins_dynamic.h [hunk 4]
```diff
@@ -36,232 +52,303 @@ std::size_t TDimension,
...
-    typedef typename TreeNodeType::SizeType             SizeType;        // std::size_t
-    typedef typename TreeNodeType::IndexType            IndexType;       // std::size_t
+    using TreeNodeType = TreeNode<Dimension, TPointType, TPointerType, TIteratorType, TDistanceIteratorType>;
 
-    typedef Tvector<CoordinateType,Dimension>           CoordinateArray;
-    typedef Tvector<SizeType,Dimension>                 SizeArray;
-    typedef Tvector<IndexType,Dimension>                IndexArray;
+    using CoordinateType = typename TreeNodeType::CoordinateType; // double
+    using SizeType = typename TreeNodeType::SizeType;             // std::size_t
+    using IndexType = typename TreeNodeType::IndexType;           // std::size_t
```
Comment (roigcarlo) [inline kratos/spatial_containers/bins_dynamic.h:91]: We should add a `using BoundingBoxType = BoundingBox<PointType>` so the user doesn't have to know the implementation while trying to call the `GetBoundingBox` he cannot use `auto`

File: kratos/spatial_containers/bins_dynamic.h (hunk 1 above)
Comment (loumalouomega) [reply kratos/spatial_containers/bins_dynamic.h]: Yes, we have. But I can revert it if it is a problem

File: kratos/spatial_containers/bins_dynamic.h (hunk 3 above)
Comment (loumalouomega) [reply kratos/spatial_containers/bins_dynamic.h]: okay

File: kratos/spatial_containers/bins_dynamic.h (hunk 2 above)
Comment (loumalouomega) [reply kratos/spatial_containers/bins_dynamic.h]: okay

File: kratos/spatial_containers/bins_dynamic.h (hunk 1 above)
Comment (roigcarlo) [reply kratos/spatial_containers/bins_dynamic.h]: Na, its just the first time i see it in Kratos S:

File: kratos/spatial_containers/bins_dynamic.h (hunk 4 above)
Comment (loumalouomega) [reply kratos/spatial_containers/bins_dynamic.h:91]: Yes

File: kratos/spatial_containers/bins_dynamic.h (hunk 3 above)
Comment (loumalouomega) [reply kratos/spatial_containers/bins_dynamic.h]: Due to the evil nature of intrussive pointers, some node constructors are removed, and I need to do some minor modification of bb class in ordetr to make it work

File: kratos/spatial_containers/bins_dynamic.h (hunk 2 above)
Comment (loumalouomega) [reply kratos/spatial_containers/bins_dynamic.h]: Again due to evil nature of intrussive pointers I cannot use this, and contrary to the point constructor it is not possible to modify the BB beacuse there is no common interface. I will keep it as it is.

**Comment by loumalouomega** [comment]
@roigcarlo I tried to applied in the most possible way your comments, but due to the evil requirements of intrussive pointers, the node lacks some constructors and therefore I needed to slightly modify the constructor of the BB and preserve current behaviour for iterators constructor due to incompatibility of iterators classes.

**Review by roigcarlo — APPROVED** [review]
Ty
